src/controller/CommissioningWindowParams.h

```diff
@@ -0,0 +1,185 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/CHIPCallback.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/NodeId.h>
+#include <lib/core/Optional.h>
+#include <lib/support/Span.h>
+#include <setup_payload/SetupPayload.h>
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace Controller {
+
+// Passing SetupPayload by value on purpose, in case a consumer decides to reuse
+// this object from inside the callback.
+typedef void (*OnOpenCommissioningWindow)(void * context, NodeId deviceId, CHIP_ERROR status, SetupPayload payload);
+typedef void (*OnOpenCommissioningWindowWithVerifier)(void * context, NodeId deviceId, CHIP_ERROR status);
+typedef void (*OnOpenBasicCommissioningWindow)(void * context, NodeId deviceId, CHIP_ERROR status);
+
+template <typename Derived>
+class CommissioningWindowCommonParams
+{
+public:
+    CommissioningWindowCommonParams() = default;
+
+    NodeId GetNodeId() const { return mNodeId; }
+    // The node identifier of device
+    Derived & SetNodeId(NodeId nodeId)
+    {
+        mNodeId = nodeId;
+        return static_cast<Derived &>(*this);
+    }
+
+    System::Clock::Seconds16 GetTimeout() const { return mTimeout; }
+    // The duration for which the commissioning window should remain open.
+    Derived & SetTimeout(System::Clock::Seconds16 timeout)
+    {
+        mTimeout = timeout;
+        return static_cast<Derived &>(*this);
+    }
+    Derived & SetTimeout(uint16_t timeoutSeconds) { return SetTimeout(System::Clock::Seconds16(timeoutSeconds)); }
+
+    // The PAKE iteration count associated with the PAKE Passcode ID and
+    // ephemeral PAKE passcode verifier to be used for this commissioning.
+    uint32_t GetIteration() const { return mIteration; }
+    Derived & SetIteration(uint32_t iteration)
+    {
+        mIteration = iteration;
+        return static_cast<Derived &>(*this);
+    }
+
+    // The long discriminator for the DNS-SD advertisement.
+    uint16_t GetDiscriminator() const { return mDiscriminator; }
+    Derived & SetDiscriminator(uint16_t discriminator)
+    {
+        mDiscriminator = discriminator;
+        return static_cast<Derived &>(*this);
+    }
+
+private:
+    NodeId mNodeId                    = 0;
+    System::Clock::Seconds16 mTimeout = System::Clock::Seconds16(300);
+    uint32_t mIteration               = 1000;
+    uint16_t mDiscriminator           = 0;
```

Comment:
A problem with the builder pattern thing is that it's not clear which things are required and which are optional....  For some cases (the timeout/iteration) we are now making them optional and defaulting. But the node id (which should be `kUndefinedNodeId`) we are just silently allowing invalid values in the callee.  And for discriminator we are just defaulting a valid value without any sort of "you have to have set this" sentinel.

What we should do if we do a builder is that anything that is not being defaulted sanely should have a sentinel value at init, and then when one tries to use the value we should fail out if the sentinels are still set.

Reply:
Totally agree, other than some helpful defaults (for iteration/timeout), I have initialized members to either sentinel value (`kUndefinedNodeId`) or made them `Optional` type when the sentinel is either not as clear or if the entire range is valid: https://github.com/project-chip/connectedhomeip/pull/34209/commits/da28dd643e7419092f7a073ab2011fda251f2504